File: src/com/huawei/codecraft/BackgroundThread.java
package com.huawei.codecraft;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.http.HttpHeaders;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.locks.ReentrantLock;

import static com.huawei.codecraft.Utils.*;

public class BackgroundThread {
    boolean m_exited = false;
    boolean m_working = false;

    static int questionId = 0;
    Thread m_thread;

    ReentrantLock m_lock = new ReentrantLock(false);
    private static final BackgroundThread obj = new BackgroundThread();

    private static CloseableHttpClient client;


    private BackgroundThread() {
    }

    public static BackgroundThread Instance() {
        return obj;
    }

    public int sendQuestion(String question) {
        Question q = new Question(questionId, question);
        m_lock.lock();
        input.add(q);
        m_lock.unlock();
        return questionId++;
    }

    public int getAnswer(int questionId) {
        assert questionId != -1;
        int ans = -1;
        m_lock.lock();
        if (output.containsKey(questionId)) {
            ans = output.get(questionId);
            output.remove(ans);
        }
        m_lock.unlock();
        return ans;
    }


    public static class Question {
        public int id;
        public String qus;

        public Question(int id, String qus) {
            this.id = id;
            this.qus = qus;
        }
    }


    public static Queue<Question> input = new ArrayDeque<>();
    public static HashMap<Integer, Integer> output = new HashMap<>();


    public void init() {
        RequestConfig.Builder requestBuilder = RequestConfig.custom()
                .setConnectTimeout(3000) // 连接超时3秒
                .setSocketTimeout(10000);  // 数据读取超时10秒
        client = HttpClients.custom().setDefaultRequestConfig(requestBuilder.build()).build();
        m_thread = new Thread(this::ThreadMain, "BackgroundThread");
        m_thread.start();
    }

    private void ThreadMain() {
        while (!m_exited) {
            boolean hasQuestion = true;
            m_lock.lock();
            if (input.isEmpty()) {
                hasQuestion = false;
            }
            m_lock.unlock();
            if (!hasQuestion) {
                // not update
                try {
                    //noinspection BusyWait
                    Thread.sleep(1);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
                continue;
            }
            m_working = true;
            m_lock.lock();
            Question question = input.poll();
            m_lock.unlock();
            //询问问题
            assert question != null;
            int ans = 0;
            try {
                ans = sendRequest(question.qus);
            } catch (Exception e) {
                printError("error in sendRequest:" + e);
            }
            m_lock.lock();
            output.put(question.id, ans);
            m_lock.unlock();
            m_working = false;
        }
    }


    @SuppressWarnings("all")
    private static int sendRequest(String question) {
        String url = "https://infer-app-modelarts-cn-southwest-2.myhuaweicloud.com/v1/infers/760efa2e-3801-4844-aa02-1dc4511248db";
        String appCode = "02073deccf6c4a6696f26d7f78c2d22a13168b61f77b45829f9d2404e32d9bfa";
        HttpPost httpPost = new HttpPost(url);
        httpPost.setHeader(HttpHeaders.CONTENT_TYPE, "application/json; charset=utf-8");
        httpPost.setHeader("X-Apig-AppCode", appCode);
        Map<String, Object> body = new HashMap<>();
        question += " Answer no more than one word";
        body.put("prompt", question);
        body.put("temperature", 0.0);
        ObjectMapper objectMapper = new ObjectMapper();
        try {
            String jsonString = objectMapper.writeValueAsString(body);
            StringEntity entity = new StringEntity(jsonString, ContentType.create("application/json", StandardCharsets.UTF_8));
            httpPost.setEntity(entity);
            CloseableHttpResponse responseBody = client.execute(httpPost);
            Map responseMap = new ObjectMapper().readValue(EntityUtils.toString(responseBody.getEntity()), Map.class);
            Object object = responseMap.get("text");
            if (object == null) {
                printError("error in send message");
                return 0;
            }
            String text = (String) object;
            printDebug("question:" + question + ",LLMAns:" + text);
            // add some additional practice here
            for (int i = 0; i < text.length(); i++) {
                if (text.charAt(i) == 'A' || text.charAt(i) == 'a') {
                    return 0;
                }
                if (text.charAt(i) == 'B' || text.charAt(i) == 'b') {
                    return 1;
                }
                if (text.charAt(i) == 'C' || text.charAt(i) == 'c') {
                    return 2;
                }
                if (text.charAt(i) == 'D' || text.charAt(i) == 'd') {
                    return 3;
                }
            }
        } catch (JsonProcessingException e) {
            printError(e.toString());
        } catch (IOException e) {
            printError(e.toString());
        }
        return 0;
    }


    void exitThread() throws InterruptedException, IOException {
        m_exited = true;
        if (m_thread != null) {
            m_thread.join();
        }
        client.close();
    }

    boolean IsWorking() {
        return m_working;
    }


}

File: BackgroundThread.cpp
/*
 * Description: 背景线程类定义，背景线程负责在后台不停的询问大模型就行，反正比去年简单就对了
 *   主线程只需要查询即可，这样可避免主线程出现跳帧的情况。
 * Date: 2024-04-18
 */

#include "BackgroundThread.h"
#include "Utils.h"

using std::make_unique;
using std::thread;

void BackgroundThread::init() {
    const std::string API_URL = "https://infer-app-modelarts-cn-southwest-2.myhuaweicloud.com/v1/infers/760efa2e-3801-4844-aa02-1dc4511248db";
    const std::string CONTENT_TYPE = "Content-Type: application/json;charset=utf-8";
    const std::string AUTH_TOKEN = "X-Apig-AppCode:02073deccf6c4a6696f26d7f78c2d22a13168b61f77b45829f9d2404e32d9bfa";
    curl = curl_easy_init();
    // url
    curl_easy_setopt(curl, CURLOPT_URL, API_URL.c_str());
    // 请求方式
    curl_easy_setopt(curl, CURLOPT_POST, 1L);

    // 设置连接超时为5秒
    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 5L);

    // 设置读取超时为10秒
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);

    // 头
    headers = curl_slist_append(headers, CONTENT_TYPE.c_str());
    headers = curl_slist_append(headers, AUTH_TOKEN.c_str());
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

    //指定证书
//    curl_easy_setopt(curl, CURLOPT_CAINFO, "./curl-ca-bundle.crt");//线上注释掉
    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0);
    //取消验证也行，线上注释掉

    // 回调
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);


    //开线程
    m_thread = make_unique<thread>(&BackgroundThread::threadMain, this);
    m_thread->detach();//这样子进程才能死掉，不让ctrl+c怎么都杀不死进程，让他自己去检测结束
}


void BackgroundThread::threadMain() {
    long long int l = runTime();
    int lastFrameId = 0;
    int count = 0;
    while (!m_exited) {
        bool hasQuestion = true;
        m_lock.lock();
        if (input.empty()) {
            hasQuestion = false;
        }
        m_lock.unlock();
        if (!hasQuestion) {
            // not update
            long long int e = runTime();
            int curFrameId = frameId;
            if (curFrameId == lastFrameId) {
                count++;
            } else {
                count = 0;
            }
            lastFrameId = curFrameId;
            if ((e - l) > 1000 * 25 * 60 || count > 1000 * 60) {
                //超过25分钟结束，60秒没改变frameId结束
                m_exited = true;
                break;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(1));
            continue;
        }
        m_working = true;
        m_lock.lock();
        Question question = input.front();
        input.pop();
        m_lock.unlock();
        //询问问题
        int ans = sendRequest(question.qus);
        m_lock.lock();
        output[question.id] = ans;
        m_lock.unlock();
        m_working = false;
    }

}


BackgroundThread &BackgroundThread::Instance() {
    static BackgroundThread obj;
    return obj;
}

size_t BackgroundThread::writeCallback(char *ptr, size_t size, size_t nmemb, std::string *data) {
    data->append(ptr, size * nmemb);
    return size * nmemb;
}

int BackgroundThread::sendRequest(string question) {
    question += " Answer no more than one word";
    //实测topk比贪心更快
    string total = R"({"prompt":")" + question + R"(","top_k":1})";
//    string total ="{\"prompt\":\""  + question +"{\",\"temperature\":0}";
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, total.c_str());
    long long int l1 = runTime();
    CURLcode res = curl_easy_perform(curl);
    long long int l2 = runTime();
    // 检查错误
    if (res != CURLE_OK) {
        printError("error in send http" + to_string(res));
        return 0;
    }
    if (response.length() <= 9 + 2) {
        printError("error in response" + response);
        return 0;
    }
    string s = response.substr(9);
    response.clear();
    s = s.substr(0, s.length() - 2);
    printError(
            "frame:" + to_string(frameId) + ",question:" + question + ",LLMAns" + s + ",runTime:" + to_string(l2 - l1));
    for (char i: s) {
        if (i == 'A' || i == 'a') {
            return 0;
        }
        if (i == 'B' || i == 'b') {
            return 1;
        }
        if (i == 'C' || i == 'c') {
            return 2;
        }
        if (i == 'D' || i == 'd') {
            return 3;
        }
    }
    return 0;
}

void BackgroundThread::exitThread() {
    m_exited = true;
//    if (m_thread) {
//        m_thread->join();//会杀不死进程
//    }
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
}

bool BackgroundThread::isWorking() const {
    return m_working;
}

int BackgroundThread::sendQuestion(string question) {
    Question q{questionId, std::move(question)};
    m_lock.lock();
    input.push(q);
    m_lock.unlock();
    return questionId++;
}

int BackgroundThread::getAnswer(int id) {
    assert (id != -1);
    int ans = -1;
    m_lock.lock();
    if (output.find(id) != output.end()) {
        ans = output[id];
        output.erase(ans);
    }
    m_lock.unlock();
    return ans;
}
